### PR TITLE
added ability to make analytics calls to an iframe

### PR DESCRIPTION
**Pull request by cory-work**
This PR implements to https://github.com/ampproject/amphtml/issues/1352

### CODE REVIEW

**Comment by cory-work** [comment]
I signed it!

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 1]
```diff
@@ -273,6 +290,81 @@ export class AmpAnalytics extends AMP.BaseElement {
       return val + argList;
     });
 
+    if (request.startsWith("iframe")) {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/amp-analytics.js]: `#startsWith` is a very new ES6 feature. Might be better to do `#indexOf == 0` or `/^iframe/.test(request)`.

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 2]
```diff
@@ -273,6 +290,81 @@ export class AmpAnalytics extends AMP.BaseElement {
...
+   * Callback for events that are registered by the config's iframe triggers.
+   * This method generates the request and sends the request out.
+   *
+   * @param {!JSONObject} request string which contains the url to be loaded
+   *                              into the iframe
+   * @param {!JSONObject} trigger JSON config block that resulted in this event.
+   * @param {!Object} unusedEvent Object with details about the event.
+   * @private
+   */
+  handleIframe_(request, trigger, unusedEvent) {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/amp-analytics.js]: `unusedEvent` is actually used. :smiley:

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 3]
```diff
@@ -273,6 +290,81 @@ export class AmpAnalytics extends AMP.BaseElement {
...
+      this.iframeQueue_.push(vars);
+    }
+
+    // has this iframe been used to analytics requests
+    let iframe = this.loadedIframes_[trigger.request];
+
+    if (!iframe) {
+      // does the iframe exist on the page somewhere else
+      iframe = document.getElementById(trigger.request);
+      if (iframe && iframe.nodeName.toLowerCase() != "iframe") {
```
Comment (jridgewell) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Not sure I follow what this code block is doing. Can you explain?

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 2 above)
Comment (cory-work) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Yeah I struggled with that. My reasoning is that the unusedEvent gets passed around in private methods and isn't actually used in any of them, so in deference to the original intent I left it as is.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 3 above)
Comment (cory-work) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Sure thing! In the original implementation all analytics calls are made by immediate requests. Using the iframe approach a race condition was created where there is a lag between the first analytics call being made and the handling iframe loading. To minimize this window I made a queue to handle all requests until the iframe is loaded. Once it is loaded the queuing is turned off and the queue is emptied.

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 2 above)
Comment (jridgewell) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Unused params are a convention so we can have type checking on the parameter without actually referencing it inside the function (which would throw a linting error). Can you change it to just `event`?

File: extensions/amp-analytics/0.1/amp-analytics.js [hunk 4]
```diff
@@ -273,6 +290,81 @@ export class AmpAnalytics extends AMP.BaseElement {
       return val + argList;
     });
 
+    if (request.indexOf("iframe") === 0) {
+      this.handleIframe_(request, trigger, unusedEvent);
```
Comment (cory-work) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Good tip...done

File: extensions/amp-analytics/0.1/amp-analytics.js (hunk 2 above)
Comment (cory-work) [inline extensions/amp-analytics/0.1/amp-analytics.js]: Ah gotcha...I'll make that change.

**Comment by R-O-P** [comment]
Hello there,
Is there any example of using the iframe for trackers?
Is this better than simply using the external config, where the external config is served from a host that has first cookies?

**Pull request closed without merging**